spec: draft

CosmicToast opened the pull request:
First specification draft.
Note that the specification is not the source of truth, and is purely here to help people implementing CNI understand it faster without learning raku.

Commit:
spec: draft 1

CosmicToast commented:
Open for comments.
CC @Johann150

Johann150 commented on SPEC.md:
```diff
@@ -0,0 +1,121 @@
+It's a key, an equals, and a value, with arbitrary amounts of spaces before and after the `=`.
+
+A value may contain any UTF-8 characters except vertical whitespace or the `#` sign (note: this is expanded upon in ini-compatibility).
+It may not start or end with whitespace, or start with a backtick.
```
So is trailing (non-vertical) white space part of the value? If so, there should be a test for that.
And what happens with he how does this interact with the flexspace extension, is it still part of the value?

CosmicToast replied:
The trailing non-vertical whitespace is not part of the value (as it may not end with whitespace).
This does not interact with flexspace (for the same reason).
If you want a value that starts/ends with whitespace, you must use raw values.

Johann150 commented on SPEC.md:
```diff
@@ -0,0 +1,121 @@
+Implementations are not required to validate it, though.
+
+### Comments
+A comment starts with a `#` (note: this is expanded upon in ini-compatibility) and ends at the next newline.
```
```suggestion
A comment starts with a `#` (note: this is expanded upon in ini-compatibility) and ends at the next vertical space.
```

CosmicToast replied:
I'll have to experiment with this in the grammar (right now a newline is the only terminator).
I agree that this makes sense though, so I'll try it and if it doesn't break anything, will change both.

formerly-a-trickster commented:
I'd like to have a well-defined resolution strategy for files that use the same key several times.

Like for example, the last definition of that key wins. So the following

```
source = main.zip
source = src.zip
```

would result in `source=src.zip`

Johann150 commented on SPEC.md:
```diff
@@ -0,0 +1,121 @@
+
+### Key
+A key is composed of alphanumeric (`[a-zA-Z0-9]`) characters, underscores, and dashes.
+Keys may contain dots, but may not start or end with them.
+Two dots in a row are disallowed (due to being nonsensical), but implementations do not need to explicitly disallow them.
```
If you would add something about the semantic meaning of the dots, the second sentence would make more sense.

CosmicToast replied:
Will be added in draft 2.

Johann150 commented on SPEC.md:
```diff
@@ -0,0 +1,121 @@
+
+### Comments
+A comment starts with a `#` (note: this is expanded upon in ini-compatibility) and ends at the next newline.
+A comment may start on its own line, at any point inside of a value, after a section declaration, or after a raw value.
```
Since values may not contain `#`, starting "inside of a value" is technically impossible. Maybe you have a better idea how to formulate this, mine would be:
```suggestion
A comment may start on its own line, at any point "inside" of a value (thus ending it), after a section declaration, or after a raw value.
```

CosmicToast replied:
LGTM, will be added in draft 2.

Johann150 commented on SPEC.md:
```diff
@@ -0,0 +1,121 @@
+It is impossible to have comments inside of a raw value, or inside of a section declaration.
+
+## Ini-Compatibility
+For compatibility with INI, the `;` character may also be used for comments.
+In this case, anytime the `#` is a significant character in the core language, the `;` character should be as well.
```
```suggestion
For compatibility with INI, the `;` character may also be used for comments, i.e. it is treated as if it were a `#`, except in raw-values.
```

CosmicToast replied:
`#` has no special meaning inside of raw values, so I'll omit that part. Besides that, will be added in draft 2.

Johann150 replied:
I only put that there since I removed the "significant character" part. But if you think its okay like that it's also fine 😄

Johann150 commented on SPEC.md:
```diff
@@ -0,0 +1,121 @@
+* Between the opening section bracket, the section name, and the closing section bracket.
+* Between the key in a key-(raw)value pair, the `=`, and the (raw)value.
+* Between a (raw) value and a comment.
+
```
```suggestion
This also means that comments may be placed inside of section headings or between keys and the `=`, if the construct's syntax is continued on the next line.

```
The phrasing is not the best with this one, but I hope you get the point.

CosmicToast replied:
No, comments may not be placed within such constructs, since comments are statements and not whitespace.
This extension is already mostly there since it's harder to disallow it than it is to let it be.
Having comments inside of flexspace may be an extension, but I don't actually expect people to ever do this as it is.

Johann150 commented on SPEC.md:
```diff
@@ -0,0 +1,121 @@
+* Between a (raw) value and a comment.
+
+### Tabulation
+In implementations with the "tabulation" extension, whitespace before and after statements is not significant.
```
This is tested in `tests/core/bareword/03.cni` too. If this ability is only added in this extension, that test should be (re)moved.

CosmicToast replied:
You're right, core/bareword/03 will be removed with draft 2.

Johann150 commented on SPEC.md:
```diff
@@ -0,0 +1,121 @@
+
+The following API endpoints provide all of this functionality:
+
+### Walk (Flat|Rec)
```
The `Flat` and `Rec` names are not very descriptive. Since we already have the trie/tree analogy, maybe these could be named something like `WalkTree` and `WalkLeaves` (adjusting the other functions appropriately).
For me, `Flat` does not really make sense here because that implies flattening the tree into a list or something. And I don't really get what `Rec` should stand for.

CosmicToast replied:
Rec was meant to stand for `Recursive`.
I'm ok with `Tree` and `Leaves`, will see if we get further comments re: the terminology.

Johann150 commented on SPEC.md:
```diff
@@ -0,0 +1,121 @@
+
+### Sub (Flat|Rec)
+SubFlat|SubRec takes a pattern, and constructs a new trie.
+The new trie will contain all of the key-value pairs that match under the corresponding `Walk*` function, but with the pattern (and the dot immediately after it) stripped.
```
Should this be `key-(raw)value` pairs?
```suggestion
The new trie will contain all of the key-(raw)value pairs that match under the corresponding `Walk*` function, but with the pattern (and the dot immediately after it) stripped.
```
I know it refers to another idea here, but still people might be confused.

Johann150 replied:
Also maybe removed instead of stripped? Stripped seems to me more of a slang word, while removed should be clearer.

CosmicToast replied:
LGTM on both, will be changed in draft 2.

Johann150 commented on SPEC.md:
````diff
@@ -0,0 +1,121 @@
+It can be implemented in pseudocode using the `Walk*` functions like so:
+
+```
+ListFlat (pattern):
+	var output_list
+	WalkFlat (pattern, (key, value){
+		output_list.append(value)
+	})
+	return output_list
+ListRec (pattern):
+	var output_list
+	WalkRec (pattern, (key, value){
+		output_list.append(value)
+	})
+	return output_list
````
The syntax of this pseudocode seems a bit strange to me, especially the anonymous function syntax. It's a bit difficult to convey an advanced concept like a closure in pseudocode, and I don't have a better idea (yet).

CosmicToast replied:
Yeah, it'd be nice to have some conventions for pseudocode, but as of right now we don't have any and this is (hopefully) clear enough.

Johann150 replied:
well we could look at what ~~Pseudocode~~ Python does
```suggestion
ListFlat (pattern):
	var output_list
	WalkFlat (pattern, lambda key, value:
		output_list.append(value)
	)
	return output_list
ListRec (pattern):
	var output_list
	WalkRec (pattern, lambda key, value:
		output_list.append(value)
	)
	return output_list
```
looks a bit more concise to me, but this much 🤏

Johann150 commented on SPEC.md:
````diff
@@ -0,0 +1,121 @@
+It can be implemented in pseudocode using the `Walk*` functions like so:
+
+```
+SubFlat (pattern):
+	var output_trie
+	WalkFlat (pattern, (key, value){
+		output_trie[key.strip-from-start(pattern + ".")] = value
+	})
+	return output_trie
+SubRec (pattern):
+	var output_trie
+	WalkRec (pattern, (key, value){
+		output_trie[key.strip-from-start(pattern + ".")] = value
+	})
+	return output_trie
````
Same comment as with `ListFlat` and `ListRec` pseudocode applies here.

Johann150 reviewed (commented):
Should we also recommend `KeyFlat` and `KeyReq` functions?

Johann150 commented:
And I forgot one more thing: For non Perl and/or `/[Rr]eg[Ee]xp?/` savvy people a list of code points for "vertical whitespace" would be nice, since vertical white space - unlike general whitespace - is not well defined by Unicode.
https://regex101.com says that `\v` in PCRE is identical to `[\x{2028}\n\r\x{000B}\f\x{2029}\x{0085}]`, so that would be these Unicode code points (with their C escape codes):

```
- U+000A Line Feed (`\n`)
- U+000B Vertical Tabulation (`\v`)
- U+000C Form Feed (`\f`)
- U+000D Carriage Return (`\r`)
- U+0085 Next Line
- U+2028 Line Separator
- U+2029 Paragraph Separator
```

CosmicToast commented:
P.S. I'm going to gather comments for a few days before writing/releasing draft 2, just to avoid having 500 total draft versions :)

Johann150 commented on SPEC.md:
```diff
@@ -0,0 +1,121 @@
+In this case, anytime the `#` is a significant character in the core language, the `;` character should be as well.
+
+## Extensions
+All extensions are fully optional.
+Currently, multiple of them are consequences of keeping the grammar lean, and are documented, but only parts of their behaviors are desirable.
```
I understand this as "fully optional for libraries to implement", while from #2 I gather it should mean fully optional to library implementers as well as library users.
```suggestion
All extensions are fully optional to be implemented. Implementations should allow for specific extensions to be en- or disabled.
Currently, multiple of them are consequences of keeping the grammar lean (so you will probably find it easier to implement them than not), and are documented, but only parts of their behaviors are desirable.
```

CosmicToast replied:
When I say "users" I mean end-users (application users). I.e the people writing the CNI files.
Library users have the option on the basis of the library they end up picking.

The intended workflow is:
1. You realize you're working with a CNI file, you know what CNI is (`core`) and can write it.
2. If the application author tells you to use an extension (or you realize that an extension is available), you know you can take advantage of its features for that application only.

Basically, `ext` is for stuff that I don't want CNI as a whole to be associated with / preassumed availability of.
`ini` can in fact be considered to be `ext/ini`, but it's separated out (right now) for convenience purposes.

Johann150 commented on SPEC.md:
```diff
@@ -0,0 +1,121 @@
+
+### Tabulation
+In implementations with the "tabulation" extension, whitespace before and after statements is not significant.
+This means you may have arbitrary whitespace in the following locations:
+* In front of a section declaration.
+* In front of a key-(raw)value pair.
+* In front of a comment.
+* Between a key-rawvalue pair and the next statement.
+* Between a section declaration and the next statement.
```
Does this mean that `core` does not allow empty lines? On the other hand, the name suggests this should only allow for arbitrary *horizontal* whitespace.
The lines between `core`, `flexspace` and `tabulation` are a bit blurry here.

CosmicToast replied:
You're right, this should be *horizontal* whitespace.
`core` allows empty lines, and in fact core tests have empty lines (this is a mild reminder that the test suite and ref. impl is still the source of truth, SPEC.md is here to help people interpret the two of them).

Commit:
spec: draft 2

Commit:
spec: add vertical whitespace definition

Commit:
spec: clarify multiple key definition resolution strategy

CosmicToast commented:
I think I went through everything here.
How's it looking now?

Johann150 commented on Parser.pm6:
```diff
@@ -65,23 +65,23 @@ class pCNI is Hash {
+		return self.SubLeaves($prefix).values.list;
 	}
 
 	# you are also recommended to implement WalkFlat and WalkRec alongside the above
```
```suggestion
	# you are also recommended to implement WalkLeaves and WalkTree alongside the above
```
I don't know if `KeyLeaves` and `KeyTree` might also be interesting.

CosmicToast replied:
Walker pattern gets called walk, surely.
It's not like it gives you a set of keys, eh?

Johann150 replied:
Yes, the walk functions stay as they are (apart from updating their names to be consistent as in my code suggestion). I meant KeyLeaves/KeyTree as *additional* API functions might be interesting, since we also have the List... functions.

CosmicToast replied:
Oh, now I see.
I think that's interesting, as well as SectionTree and SectionLeaves.
I don't want to require too much from library writers though.
Then again, the API is purely a suggestion / informative...

Johann150 commented on SPEC.md:
```diff
@@ -0,0 +1,150 @@
+
+### Tabulation
+In implementations with the "tabulation" extension, whitespace before and after statements is not significant.
+This means you may have arbitrary horizontal whitespace in the following locations:
+* In front of a section declaration.
+* In front of a key-(raw)value pair.
+* In front of a comment.
+* Between a key-rawvalue pair and the next statement.
+* Between a section declaration and the next statement.
```
Couldn't this be simplified to "at the start of any line"?

CosmicToast replied:
The last two translate to something beyond that. The first 3 could, though, I'll think about it.

Johann150 commented on SPEC.md:
```diff
@@ -0,0 +1,150 @@
+This means that you may have arbitrary whitespace (not only spaces, but also tabs, newlines, zero-width spaces, and so on) in the following locations:
+* Between the opening section bracket, the section name, and the closing section bracket.
+* Between the key in a key-(raw)value pair, the `=`, and the (raw)value.
+* Between a (raw) value and a comment.
```
Horizontal whitespace between values and comments is already allowed by `core`, as are empty lines. Is this rule still necessary?

CosmicToast replied:
I think it's fine to keep it for clarity, since this is a companion document.

Johann150 commented on SPEC.md:
```diff
@@ -0,0 +1,150 @@
+
+### Key-(Raw)Value Pairs
+A key-value and key-rawvalue pair look very similar.
+It's a key, an equals, and a value, with arbitrary amounts of spaces before and after the `=`.
```
I think to be consistent we should allow horizontal whitespace instead of just spaces.
```suggestion
It's a key, an equals, and a value, with arbitrary amounts of horizontal whitespace before and after the `=`.
```

CosmicToast replied:
I think core should only require people to check for ' ' as spacings, to simplify some parsers.
So if anything I would rather reverse it (if they do non-significant whitespace properly they end up auto-implementing the flexspace and tabulation extensions, eh?)
OTOH, users might want to be able to insert hard tabs.
I'll have to think about this.
Might end up removing tabulation/flexspace and adding those into core tbh.

Johann150 reviewed (commented):
I hope you'll not beat me with a stick for this since you are pushing for the Raku grammar to be the sole specification, but I made some EBNF syntax files for CNI. They helped in noticing some odd things to annoy you with. :P
https://github.com/Johann150/cni-format/tree/ebnf/ebnf

CosmicToast commented:
I'll think about all of these and do a draft 3 this weekend.
Re: EBNF - I don't mind it existing, I just don't want it to be "official".
I encourage you to add it to the project's wiki though (with the obvious disclaimers)!

Johann150 commented:
If you mean the literal wiki, that won't be possible since I do not have permissions to edit the wiki. I would rather make a pull request to add the files into the repository after we are finished here and I've updated the files accordingly.

CosmicToast commented:
I do mean the literal wiki.
Looks like it had some restrictions on it - I've removed them.
Like I said, I don't want it to be official, and I consider everything into the repo proper to be official, thus the conundrum :slightly_smiling_face: 
I'm fine with as many informative/unofficial things as people want, just not the repo proper - and that's exactly what the attached wikis are for.

CosmicToast commented:
I've decided to integrate flexspace and tabulation into core cni.
The reason I've excluded them so far is because they allow users to do error prone things.
For example:
```
[
    this.is.valid
]
[ # this is not valid
     neither # is this
] # but this is
```

Instead, I'm simply going to have a mention in the SPEC that some technically valid syntax is discouraged, and point people to the testing suite.

Either way, this means I need to rewrite a good chunk of the tests, so this is why things are taking a while, sorry ^^;;

Commit:
draft 3

Real life has really been getting in the way, so I'm taking a little bit of time now to just pushing something out.
Hopefully future drafts will require less modification, and thus be more plausible to just do in one shot (polishing can be done later).

Summary:
* flexspace and tabulation are no longer extensions, they're built-in
* some failing tests that may eventually be accepted as extensions are removed
  a better mechanism for this might be preferable
* finishing renaming some APIs
* add some complements to the recommended APIs
* add some notes about encouragements etc

CosmicToast commented:
Sorry this took a bit.
It's a relatively large-ish update, and real life has been getting in the way of work recently.
It still is, mind you, but hopefully with smaller future iterations, more can get done more often.

Johann150 commented on SPEC.md:
```diff
@@ -132,10 +128,25 @@ SubTree (pattern):
+### Section (Leaves|Tree)
+SectionLeaves|SectionTree takes a pattern and returns a set of prefixes.
+It can be calculated roughly as follows:
+1. Get the corresponding list of KeyLeaves|KeyTree.
+2. For each key:
+	1. If the key does not contain a `.`, continue.
+	2. Remove the final `.` and all text that follows from the key.
+	3. Add the result to the result set.
+	4. Add the result to the list of keys to process.
+3. Return the resulting set.
```
I'm not sure my alternate phrasing for 2.2 makes it clearer or more confusing, but here it goes:

```suggestion
1. Get the corresponding set from KeyLeaves|KeyTree.
2. For each key in the set:
	1. If the key does not contain a `.`, continue.
	2. Remove the last section of the key (i.e. what would be matched by the PCRE `/\.[^.]+$/m`).
	3. Add the result to the result set.
	4. Add the result to the set of keys to process.
3. Return the resulting set.
````

CosmicToast replied:
You're right re: list -> set.
Re: 2.2, I think a better approach would be to simply give an example.
Will try and do that soon.
Also, should probably make KeyLeaves|Tree into a set as well.

Commit:
spec: draft 3.1

Set-ify Key(L|T) and Section(L|T).
Add an example to the Section algorithm.

CosmicToast commented:
leaving the branch up in case something comes up that required quickly reverting and working on it more
